Add MultTranspose for DSmoother and GSSmoother

pazner opened the pull request:
- For non-symmetric matrices, this requires forming the transpose matrix.
- For l1-Jacobi and lumped Jacobi, we don't implement the transpose action for non-symmetric matrices.

This allows DSmoother and GSSmoother to be used in Multigrid, which requires the transpose action for the post-smoothers.
<!--GHEX{"author":"pazner","assignment":"2025-11-14T00:46:53","editor":"tzanio","id":5111,"reviewers":["dylan-copeland","psocratis"],"merge":"2025-12-29T21:36:11","approval":"2025-12-25T13:13:45"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5111](https://github.com/mfem/mfem/pull/5111) | @pazner | @tzanio | @dylan-copeland + @psocratis | 11/13/25 | 12/25/25 | 12/29/25 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

pazner committed:
Add MultTranspose for DSmoother and GSSmoother

For non-symmetric matrices, this requires forming the transpose matrix.

For l1-Jacobi and lumped Jacobi, we don't implement the transpose action for
non-symmetric matrices.

dylan-copeland commented on linalg/sparsesmoothers.cpp:
```diff
@@ -90,22 +130,35 @@ void DSmoother::Mult(const Vector &x, Vector &y) const
+
+void DSmoother::MultTranspose(const Vector &x, Vector &y) const
+{
+   EnsureTranspose();
```
In the case of `!iterative_mode && type == 0 && iterations == 1`, when only `DiagScale` is called, can `oper` be used instead of copying the matrix entries to `At`?

pazner replied:
Yes, in fact this is true even for `type != 0`. Addressed in c95ab91de18f6a3f01cb7df1cbb6fa3ce477aab4.

pazner committed:
Special case for DSmoother::MultTranspose

Don't need to form the transpose when only the action of D (and not A) is
needed.

pazner committed:
Improve Doxygen for DSmoother and GSSmoother

pazner commented on linalg/sparsesmoothers.hpp:
```diff
@@ -63,6 +74,9 @@ class DSmoother : public SparseSmoother
+
 public:
    /// Create Jacobi smoother.
    DSmoother(int t = 0, real_t s = 1., int it = 1)
```
Doxygen improved in 3bdcae8928f886059a0c01e0c709779ed2c91638.

pazner commented on linalg/sparsesmoothers.hpp:
```diff
@@ -76,6 +90,9 @@ class DSmoother : public SparseSmoother
    /// Matrix vector multiplication with Jacobi smoother.
    void Mult(const Vector &x, Vector &y) const override;
+
+   /// Matrix vector multiplication with the transpose of the Jacobi smoother.
```
All good points. The Doxygen is improved in 3bdcae8928f886059a0c01e0c709779ed2c91638.

pazner committed:
Add old constructors for backwards compatibility

There is no implicit conversion from int to enum types

pazner committed:
constexpr

dylan-copeland commented on linalg/sparsesmoothers.cpp:
```diff
@@ -23,15 +23,31 @@ namespace mfem
+{
+   if (oper_T) { return; }
+
+   const real_t tol = 1e-14;
```
Should this `tol` be a class member with this default value and a function to set it?

pazner committed:
Use relative tolerance in SparseSmoother::EnsureTranspose

pazner committed:
Merge remote-tracking branch 'origin/master' into sparse-smoothers-transpose

# Conflicts:
#	tests/unit/CMakeLists.txt

tzanio commented:
Merged in `next` for testing...

dylan-copeland commented on linalg/sparsesmoothers.hpp:
```diff
@@ -15,67 +15,159 @@
 class GSSmoother : public SparseSmoother
 {
+public:
+   enum GSType
```
@pazner this ended up conflicting with a name in an application that uses MFEM, so I am proposing to change it in PR 5413. Using the prefix `GS_SYMMETRIC` is sufficient for this particular conflict, but I also changed it to `enum class` to be more generally safe.
